### PR TITLE
fix: Use Karras sigma schedule for Cosmos Predict2.5 SFT inference

### LINKED ISSUE #18: [Bug] Cosmos Predict2.5 inference quality mismatch with official codebase — wrong sigma schedule
# Cosmos Predict2 inference produces degraded results compared to official Cosmos codebase

## Summary

When running Cosmos-Predict2.5-2B video2world inference with FastGen using the same model checkpoint, text prompt, input image, resolution, and step count, the generated video quality is noticeably worse than the official [cosmos-predict2.5](https://github.com/nvidia-cosmos/cosmos-predict2.5) inference code. After comparing the two codebases, I identified three misalignments in the `sample()` method of `CosmosPredict2DiT` (`fastgen/networks/cosmos_predict2/network.py`).

## Reproduction

```bash
python scripts/inference/video_model_inference.py \
    --config fastgen/configs/experiments/CosmosPredict2/config_sft.py \
    --do_student_sampling False --num_steps 35 --fps 24 \
    --neg_prompt_file scripts/inference/prompts/negative_prompt_cosmos.txt \
    --input_image_file scripts/inference/prompts/source_image_paths.txt \
    --num_conditioning_frames 1 \
    - model.guidance_scale=5.0 model.net.is_video2world=True model.input_shape="[16, 21, 60, 104]"
```

## Issue 1: Wrong sigma/timestep schedule

**The most impactful difference.** The official Cosmos uses a Karras sigma schedule, while FastGen uses the default diffusers flow-shift linear schedule.

Official Cosmos (`FlowUniPCMultistepScheduler.set_timesteps` with `use_kerras_sigma=True`):
```python
sigma_max = 200
sigma_min = 0.01
rho = 7
sigmas = np.arange(num_steps + 1) / num_steps
min_inv_rho = sigma_min ** (1 / rho)
max_inv_rho = sigma_max ** (1 / rho)
sigmas = (max_inv_rho + sigmas * (min_inv_rho - max_inv_rho)) ** rho
sigmas = sigmas / (1 + sigmas)
timesteps = sigmas * 1000  # cast to int64 (truncation, not rounding)
sigmas = np.concatenate([sigmas, [0.0]]).astype(np.float32)
```

FastGen (`network.py` line ~1153):
```python
self.sample_scheduler = UniPCMultistepScheduler(
    num_train_timesteps=1000,
    prediction_type="flow_prediction",
    use_flow_sigmas=True,
    flow_shift=shift,
)
self.sample_scheduler.set_timesteps(num_inference_steps=num_steps, device=noise.device)
```
This internally computes:
```python
sigmas = np.linspace(1.0, 1/N, N)
sigmas = shift * sigmas / (1 + (shift - 1) * sigmas)  # shift=5.0
```

When I use the Karras sigma schedule in fastgen repo, the performance improves a lot but is still not as good as the cosmos original results.

**So do you have any idea what other factors would cause the degradation?**

## Issue 2: Misleading example command in scripts/inference/video_model_inference.py
In the scripts/inference/video_model_inference.py, there exists an example command which not indicate the neg_prompt_file thus the code would import another neg_prompt_file rather than the neg_prompt_file for the cosmos model, you might also want to fix this :)
```bash
# Video2World: Cosmos Predict2
    PYTHONPATH=$(pwd) FASTGEN_OUTPUT_ROOT='FASTGEN_OUTPUT' torchrun --nproc_per_node=1 --standalone \\
        scripts/inference/video_model_inference.py --do_student_sampling False \\
        --input_image_file scripts/inference/prompts/source_image_paths.txt --num_conditioning_frames 1 \\
        --config fastgen/configs/experiments/CosmosPredict2/config_sft.py \\
        - trainer.seed=1 trainer.ddp=True model.guidance_scale=5.0 model.net.is_video2world=True \\
        log_config.name=cosmos_v2w_inference
```

## Suggested Fix

In `fastgen/networks/cosmos_predict2/network.py`, the `sample()` method should:

1. Use the Karras sigma schedule matching the official Cosmos `set_timesteps` with `use_kerras_sigma=True`

```python
# Initialize internal state
self.sample_scheduler.set_timesteps(num_inference_steps=num_steps, device=noise.device)

# 2. Override with Karras sigma schedule
import numpy as np
sigma_max, sigma_min, rho = 200, 0.01, 7
sigmas = np.arange(num_steps + 1) / num_steps
min_inv_rho = sigma_min ** (1 / rho)
max_inv_rho = sigma_max ** (1 / rho)
sigmas = (max_inv_rho + sigmas * (min_inv_rho - max_inv_rho)) ** rho
sigmas = sigmas / (1 + sigmas)
timesteps_np = sigmas * self.sample_scheduler.config.num_train_timesteps
sigmas = np.concatenate([sigmas, [0.0]]).astype(np.float32)
self.sample_scheduler.sigmas = torch.from_numpy(sigmas).to("cpu")
self.sample_scheduler.timesteps = torch.from_numpy(timesteps_np).to(device=noise.device, dtype=torch.int64)
self.sample_scheduler.num_inference_steps = len(timesteps_np)
self.sample_scheduler.model_outputs = [None] * self.sample_scheduler.config.solver_order
self.sample_scheduler.lower_order_nums = 0
self.sample_scheduler.last_sample = None
self.sample_scheduler._step_index = None
self.sample_scheduler._begin_index = None
```

2. align those example commands in your scripts with those in the [readme](https://github.com/NVlabs/FastGen/blob/main/scripts/README.md)

### PR DESCRIPTION
**Pull request by mvanhorn**
## Summary
Cosmos-Predict2.5-2B video2world inference under FastGen produces noticeably degraded output compared to the official cosmos-predict2.5 codebase when using the same checkpoint, prompt, image, resolution, and step count. The reporter traced the primary cause to a wrong sigma/timestep schedule: the official Cosmos uses a Karras sigma schedule (sigma_max=200, sigma_min=0.01, rho=7) in its FlowUniPCMultistepScheduler.set_timesteps(use_kerras_sigma=True), while FastGen's CosmosPredict2DiT.sample() relies on the default diffusers flow-shift linear schedule from UniPCMultistepScheduler.

## Changes
In fastgen/networks/cosmos_predict2/network.py, modify the sample() method (around lines 1148-1161) so that, for the Cosmos SFT sampling path, after constructing the UniPCMultistepScheduler and calling set_timesteps, it overrides sample_scheduler.sigmas and sample_scheduler.timesteps with a Karras sigma schedule (sigma_max=200, sigma_min=0.01, rho=7) exactly as the official Cosmos does, including the sigmas/(1+sigmas) reparameterization, int64 timestep truncation, and resetting the scheduler's internal solver state (model_outputs, lower_order_nums, last_sample, _step_index, _begin_index). Gate this behind a sensible default or config flag so it only affects the SFT inference path the maintainer called out, not distillation. Update the inline example command(s) in scripts/inference/video_model_inference.py to pass the Cosmos negative-prompt file and otherwise align them with scripts/README.md. Add a focused unit test in tests/test_network.py asserting the produced sigma/timestep schedule matches the reference Karras values.

Fixes #18

## Summary by CodeRabbit

* **New Features**
  * Added an optional Karras sigma schedule for Cosmos Predict2 sampling, improving UniPC inference behavior and allowing the schedule to be enabled or disabled.
* **Bug Fixes**
  * Sampling now resets scheduler state before generation to avoid stale values affecting results.
  * Progress reporting now matches the actual number of timesteps.
* **Documentation**
  * Updated inference examples with clearer, more consistent command-line flags and model settings.
* **Tests**
  * Added coverage for the new schedule behavior, scheduler reset handling, and updated example commands.

### CODE REVIEW

**Comment by juliusberner** [comment]
@coderabbitai

**Comment by juliusberner** [comment]
Hi @mvanhorn,

Thanks a lot for the fix! Is there a reason why we can't just use `use_karras_sigmas` in our `FlowUniPCMultistepScheduler`?

File: fastgen/networks/cosmos_predict2/network.py [hunk 1]
```diff
@@ -767,6 +816,7 @@ def __init__(
         use_wan_fp32_strategy: bool = True,
         # FPS for temporal position embeddings
         fps: float = 24.0,
+        use_karras_sigma_schedule: bool = True,
```
[the remark this answers is not in the document]
Comment (juliusberner) [reply fastgen/networks/cosmos_predict2/network.py]: Check again

File: fastgen/networks/cosmos_predict2/network.py [hunk 2]
```diff
@@ -51,6 +51,55 @@
...
+    device: Union[torch.device, str],
+) -> None:
+    sigmas, timesteps = _build_cosmos_predict2_karras_schedule(
+        num_steps=num_steps,
+        num_train_timesteps=scheduler.config.num_train_timesteps,
+        device=device,
+    )
+    scheduler.sigmas = sigmas
+    scheduler.timesteps = timesteps
+    scheduler.num_inference_steps = len(timesteps)
```
[the remark this answers is not in the document]
Comment (juliusberner) [reply fastgen/networks/cosmos_predict2/network.py]: Check again

File: fastgen/networks/cosmos_predict2/network.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (juliusberner) [reply fastgen/networks/cosmos_predict2/network.py]: Check again @coderabbitai

File: fastgen/networks/cosmos_predict2/network.py [hunk 3]
```diff
@@ -51,6 +51,55 @@
...
+        raise ValueError(f"num_steps must be positive, got {num_steps}")
+
+    ramp = torch.arange(num_steps + 1, dtype=torch.float64) / num_steps
+    min_inv_rho = _COSMOS_KARRAS_SIGMA_MIN ** (1 / _COSMOS_KARRAS_RHO)
+    max_inv_rho = _COSMOS_KARRAS_SIGMA_MAX ** (1 / _COSMOS_KARRAS_RHO)
+    sigmas = (max_inv_rho + ramp * (min_inv_rho - max_inv_rho)) ** _COSMOS_KARRAS_RHO
+    sigmas = sigmas / (1 + sigmas)
+
+    timesteps = (sigmas * num_train_timesteps).to(device=device, dtype=torch.int64)
+    sigmas = torch.cat([sigmas, sigmas.new_zeros(1)]).to(dtype=torch.float32)
```
[the remark this answers is not in the document]
Comment (juliusberner) [reply fastgen/networks/cosmos_predict2/network.py]: Check again

File: fastgen/networks/cosmos_predict2/network.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (juliusberner) [reply fastgen/networks/cosmos_predict2/network.py]: @coderabbitai But all configs should use this scheduler?

**Review by juliusberner — APPROVED** [review]

**Comment by mvanhorn** [comment]
Really appreciate the merge on the Karras sigma schedule fix. Good to have Cosmos Predict2.5 inference matching the official codebase again, thanks.

**Comment by juliusberner** [comment]
> Really appreciate the merge on the Karras sigma schedule fix. Good to have Cosmos Predict2.5 inference matching the official codebase again, thanks.

No worries, thanks a lot for the PR!!
